### PR TITLE
wso2-jenkins-jacoco-plugin

### PR DESCRIPTION
**Pull request by hasithajayasundara**
## Purpose
The purpose of this pr is to add wso2-jenkins-jacoco plugin to build-automation-artifacts 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
+package hudson.plugins.jacoco;
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java:19]: Missing license.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import org.apache.commons.io.FileUtils;
+import org.jacoco.core.tools.ExecFileLoader;
+
+import java.io.*;
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: We do not use star imports..

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+/**
+ * Collect Jacoco resources files and create a zip file
+ *
+ * @author Hasitha Jayasundara
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: author tag is not used in wso2 convention.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    private static final String CLASS_DIRECTORY = "classes";
+    private static final String SOURCE_DIRECTORY = "sources";
+    private static final String JACOCO_RESOURCES = "jacocoResources";
+    private static final String RESOURCE_ZIP_FILE_NAME = "resources.zip";
+    private static final String JACOCO_EXEC_FILE = "jacoco.exec";
+    private static final int BYTE_BUFFER_LENGTH = 1024;
+
+    /**
+     * This method collect jacoco source,class files and
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: End docs with a dot.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    public void collectResourceFiles() {
+
+        final PrintStream logger = taskListener.getLogger();
+
+        if (isCreateZipFiles) {
+            logger.println("[JaCoCo plugin] Collecting .java , .class and jacoco.exec files");
+            final List<File> jacocoExecFiles = action.getJacocoReport().getExecFiles();
+            int fileSize = jacocoExecFiles.size();
+            if (fileSize == 0) {
+                logger.println("[JaCoCo plugin] No jacoco.exec file recorded");
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: Usually a log file without any parameters/context is a sign of code smell.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                        ExecFileLoader loader = new ExecFileLoader();
+                        for (File exec : jacocoExecFiles) {
+                            loader.load(exec);
+                        }
+                        logger.println("[JaCoCo plugin] Fuck Successfully merged jacoco.exec files");
+                        loader.save(new File(jacocoResourcesFolder + File.separator + JACOCO_EXEC_FILE),
+                                false);
+                    }
+                    logger.println("[JaCoCo plugin] Compressing Jacoco resource folder...");
+                    compressAndCopy(jacocoResourcesFolder, jobInfoDirectory + File.separator +
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: What's the reason for doing this?

File: wso2-jenkins-jacoco-plugin/src/main/resources/hudson/plugins/jacoco/JacocoPublisher/config.jelly [hunk 7]
```diff
@@ -181,5 +181,10 @@
...
   </f:entry>
-
-</j:jelly>
+  <f:entry field="createZipFiles">
+      <f:checkbox checked="${instance.selected}" title="Save coverage data"/>
+  </f:entry>
+  <f:entry>
+    <f:textbox field="jacocoResourcePath" width="78%" placeholder="Folder path to store jacoco resources" />
+  </f:entry>
+</j:jelly>
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/resources/hudson/plugins/jacoco/JacocoPublisher/config.jelly]: EOF should have a new-line.

File: wso2-jenkins-jacoco-plugin/src/main/resources/hudson/plugins/jacoco/report/CoverageReport/index.jelly [hunk 8]
```diff
@@ -19,4 +18,4 @@
       <e:breakdownTable />
     </l:main-panel>
   </l:layout>
-</j:jelly>
+</j:jelly>
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/resources/hudson/plugins/jacoco/report/CoverageReport/index.jelly]: Same as above. Please check all files you modified.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java (hunk 1 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java:19]: Added license header

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java (hunk 2 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: Removed star imports.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java (hunk 5 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: final PrintStream logger = taskListener.getLogger();

logger is taken from a TaskListener of existing implementation.Can you suggest a solution?

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java (hunk 6 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: To calculate the line coverage we need to collect all the jacoco exec files,sources and class files and create a zip file.This creates a zip file inside the build folder containing all the above mentioned files.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java (hunk 5 above)
Comment (kasunbg) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/ResourceFilesCollector.java]: Using the taskListener logger is fine. My question was on the log message that you were using. 
See #4 of http://www.masterzen.fr/2013/01/13/the-10-commandments-of-logging/

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java [hunk 9]
```diff
@@ -1,114 +1,109 @@
...
+
+    private CoverageReport(JacocoBuildAction action) {
+        this.action = action;
+        setName("Jacoco");
+    }
+
+    //	private String instructionColor;
 //	private String classColor;
 //	private String branchColor;
 //	private String complexityColor;
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java]: commented code lines... Better to remove these.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java [hunk 10]
```diff
@@ -1,114 +1,109 @@
...
+                        packageReport.add(classReport);
+                    }
+
+                    this.add(packageReport);
+                }
+            }
+            action.getLogger().println("[JaCoCo plugin] Done.");
+
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java:100]: 1. Unless there's a better reason, catching generic Exception is not a good idea.

2. Use of printStackTrace is considered a bad practice.

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java [hunk 11]
```diff
@@ -175,85 +168,107 @@ private String getMethodName(IClassCoverage classCov, IMethodCoverage methodCov)
...
+//
+//        switch (files.size()) {
+//        case 0:
+//            return HttpResponses.error(404, "No jacoco.exec file recorded");
+//        case 1:
+//            return HttpResponses.staticResource(files.get(0));
+//        default:
+//            // TODO: perhaps we want to cache the merged result?
+//            return new HttpResponse() {
+//                public void generateResponse(StaplerRequest req, StaplerResponse rsp, Object node) throws IOException, ServletException {
```
Comment (kasunbg) [inline wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java:233]: commented code.

File: wso2-jenkins-jacoco-plugin/src/main/resources/hudson/plugins/jacoco/JacocoPublisher/config.jelly (hunk 7 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/resources/hudson/plugins/jacoco/JacocoPublisher/config.jelly]: Added new line

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java (hunk 10 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java:100]: This is original source code

File: wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java (hunk 11 above)
Comment (hasithajayasundara) [reply wso2-jenkins-jacoco-plugin/src/main/java/hudson/plugins/jacoco/report/CoverageReport.java:233]: This is original source code

**Review by kasunbg — APPROVED** [review]

**Pull request merged**